### ISSUE TITLE
Remote task execution API

### ISSUE THREAD

**Issue by DrPyser** [spotify/luigi#2521]
Hi. 

Basically, I want to be able to interact with a luigi scheduler remotely. It's my understanding that luigi uses an API server that I might be able to use, but I haven't found documentation on that API, other than for the task history feature.

My goal is to deploy multiple task schedulers, in docker containers, where some tasks in one container might depend on tasks defined in another container. Those requirements on remote tasks would be managed by a task communicating with the remote schedulers, for execution and target check, etc. 

So basically, I'd need to be able to talk to schedulers, tell them to run a task(if it hasn't ran already according to their target), and then when that remote execution finishes(successfully or not), the local task would commit a target. Is it possible currently with the `luigi.server` implementation? 

I know I could always implement my own api server which interfaces [`luigi.interface`](https://luigi.readthedocs.io/en/stable/api/luigi.interface.html) if that's not the case.

Thanks!

**Comment by Tarrasch** [comment]
Can't you achieve all of this using 1 central luigi server and depend on ExternalTasks to set the task dependency boundaries between containers?

**Comment by DrPyser** [comment]
I did consider using one central scheduler. I didn't think about using ExternalTask though. One of the goal is to keep each task service as autonomous and independent as possible. In that spirit, they would each have their own database for storing targets.

That being said, yes, that approach might be the simplest one. I'll give it some more thoughts.

Thanks!

**Comment by DrPyser** [comment]
I guess an issue with using `ExternalTask`, is that this prevents the scheduler from being able to trigger the external dependencies if they haven't successfully ran yet?

**Comment by DrPyser** [comment]
I realized I didn't fully understand the role and the behavior of the scheduler. I understand better now that the scheduler doesn't have any role in trigger tasks at any point, so it's really on the worker that a remote triggering API makes any sense. What confused me, is that tasks *are* triggered directly by luigi, when they are dependencies for tasks that has been triggered externally. So the initial trigger is always external, but Luigi still has the capability/responsibility to trigger tasks as part of a dependency cascade.

At any rate, I'm implementing a websocket api over `luigi.interface` for my purpose, so I'll close this issue.

**Comment by raulsperoni** [comment]
Hi @DrPyser just stumbled with this issue while googling. I'm trying to run luigi.build from a Rest api, the idea behind it is to schedule pipelines in a web interface. Is it what you was trying to achieve? Did you succeed?

thank you.

**Comment by DrPyser** [comment]
@raulsperoni Yes, I succeeded. I'm using a websocket RPC API that uses `interface.build` to run tasks.

Right now the code is not open sourced, unfortunately. If you have any question, I can try to answer.

**Comment by raulsperoni** [comment]
Well good! I'm not familiar with websocket RPC but if you could point me in the right direction it would be great. 

I'm trying to receive via web API some task definition to run (web interface). 
I want to save that definition in a database and later check for luigi events (central_scheduler) and get back to the user immediately. 

If i'm not mistaken i need some process to pick up that task and run it, that needs to be some other process and not de api server right? So from what i understand there is no way to make luigi.build not blocking....but maybe i'm missing something? Than you.

**Comment by DrPyser** [comment]
Yes, I run the call to `luigi.build` in a separate process using [a subprocess pool](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ProcessPoolExecutor).

How do you plan on "getting back to the user"? I assume the initial http call is asynchronous, and you send back a response(e.g. a [`202` status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/202)) as soon as the task is scheduled in a separate thread?

Using a websocket allows for the call to take a long time without worrying about timeouts, while still maintaining simple request-response interactions.  

You could also use an http-based asynchronous pattern, like long-polling or streaming(see https://codeburst.io/polling-vs-sse-vs-websocket-how-to-choose-the-right-one-1859e4e13bd9).

**Comment by tashrifbillah** [comment]
Anyone else got a more descriptive solution?

**Comment by PeteW** [comment]
I'm interested in designing a more formal interface for exposing `luigi.build` over an API.

On a high level the API would require:
1. Capability to display hypothetical workflow execution plan for any workflow before running it
2. Capability to spawn workflows with roughly the same expressiveness as the command line interface
3. Capability to monitor the status and logged output of spawned workflows

Often times before I run a workflow, I want to run external setup/teardown code. Examples: getting session credentials, or, loading virtual environments before calling luigi. Airflow's design supports API-based workflow invocation but the executor is tightly coupled to the worker making virtual environments and setup/teardown poor at best. So how does one build it better? One possibility would be to develop a "execution specs" containing:
* the name of the workflow i want to run
* the path of the workflow module
* a virtual environment i want to load before running
* additional setup/teardown commands

The API would receive "execution specs", tag the request with a unique identifier, then spawn execution in some sort of forked process, likely a subprocess. The unique identifier would be returned to the user for tracking. The output from the subprocess would be captured to temp-file ostensibly under a sub-directory named with the unique id. Calls to monitor the process would return a file under the directory.

This is somewhat crude/basic design to start the conversation but based on my experience it would be effective.
A few additional thoughts:
* luigid already hosts an API. Ideally we'd tack this into the existing host
* setup/teardown commands are important. but specifying them from a client is tantamount to remote code execution which isnt very cool (or very cool depending on what kind of person you are)
